torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1598,55 +1600,39 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::pointToPoint(
   return work;
 }
 
-std::vector<at::Tensor> copyCuda(const std::vector<at::Tensor>& inputs) {
+std::vector<at::Tensor> extractCudaTensors(const std::vector<at::Tensor>& inputs) {
   std::vector<at::Tensor> cudaInputs;
   cudaInputs.reserve(inputs.size());
   for (auto& input : inputs) {
-    cudaInputs.push_back(input.to(c10::Device(c10::kCUDA, input.device().index())));
+    cudaInputs.push_back(
+        dynamic_cast<torch_lazy_tensors::compiler::TSData*>(torch::lazy::GetLtcTensor(input).GetDataHandle().get())->data());
```

Comment:
hmm, there are a couple of things that look questionable to me, with doing the data pointer access this way.

1. its TS specific, so we'd definitely want to structure it more as an 'API' we call from DDP to inside LTC, that way inside LTC we can handle either TS/XLA case differently and not expose the backend difference up to DDP layer
2. I think both GetLtcTensor and GetDataHandle are 'unsafe' in some sense, they could fail in general.  Do you have reasoning that both will always return true?  e.g. you did something that makes sure 'input' is a materialized lazytensor already?

Reply:
1. That's why I didn't make it to a LT API. The final solution is to lower the comm ops to the graph, which we rely on the dispatcher to give us the comm ops. Therefore, we shouldn't touch any DDP things at all in the final solution.
2. The caller of the function has a check to make sure they are lazy tensors. Also, `GetLtcTensor` has a release check to make sure the passed in tensors are lazy tensors. We won't have any silent failures. For `GetDataHandle`, I just looked at the implementation. I'm not exactly sure how the TORCH_CHECK will be triggered. Again, we shouldn't have any silent failures.

Reply:
ahh ok i confused GetDataHandle with CurrentDataHandle.

If you called CurrentDataHandle, it would be concerning as there could be cases where the tensor had no data_handle, and it would simply fail.

GetDataHandle otoh will always succeed, but sometimes be fast (if the tensor has already been materialized) and sometimes be slow (if it has to trigger a graph sync and materialize the tensor).

If you are not manually syncing the tensor some other way, then i expect you are needing to sync it at 'GetDataHandle' anyways.  This is safe.

It may also be as optimal as you can make it, unless you cause more syncs than you need to somehow.

Reply:
You mean materializing the tensors before the `GetDataHandle`. How would that be faster than letting `GetDataHandle` materialize the tensors for me?

Reply:
> How would that be faster than letting GetDataHandle materialize the tensors for me?

I think @wconstab just means that sync is unavoidable no matter when it happens. Either `GetDataHandle` triggers it or you could trigger it manually earlier and use `CurrentDataHandle`.

Reply:
@alanwaketan fyi i ran into some compile errors here
```
../torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp: In function ‘std::vector<at::Tensor> c10d::extractCudaTensors(const std::vector<at::Tensor>&)’:
../torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1609:94: error: ‘using LazyTensorPtr = class c10::intrusive_ptr<torch::lazy::LazyTensor> {aka class c10::intrusive_ptr<torch::
lazy::LazyTensor>}’ has no member named ‘GetDataHandle’
         dynamic_cast<torch_lazy_tensors::compiler::TSData*>(torch::lazy::GetLtcTensor(input).GetDataHandle().get())->data());
```

Reply:
oh this probably didn't get rebased correctly,

GetLtcTensor(input).GetDataHandle()

should change to 

GetLtcTensor(input)->GetDataHandle()

Reply:
I will push a fix